sycl/plugins/cuda/pi_cuda.cpp

```diff
@@ -2863,6 +2871,130 @@ pi_result cuda_piEnqueueMemUnmap(pi_queue command_queue, pi_mem memobj,
   return ret_err;
 }
 
+// USM
+pi_result cuda_piextUSMHostAlloc(void **result_ptr, pi_context context,
+                                 pi_usm_mem_properties *properties, size_t size,
+                                 pi_uint32 alignment) {
+  PI_CHECK_ERROR(cuMemAllocHost(result_ptr, size));
```

Comment:
Should there be an assert for properties to be nullptr since its not implemented?

Reply:
Alternatively, you could make this cuMemHostAlloc instead. We could map later the `CL_MEM_ALLOC_WRITE_COMBINED_INTEL` flag to `CU_MEMHOSTALLOC_WRITECOMBINED` assuming they are equivalent enough (they might not be)

Reply:
[cl_intel_unified_shared_memory.asciidoc](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/USM/cl_intel_unified_shared_memory.asciidoc) says
> `CL_MEM_ALLOC_WRITE_COMBINED_INTEL`: Request write combined (WC) memory. Write combined memory may improve performance in some cases, however write combined memory must be used with care since it may hurt performance in other cases or use different coherency protocols than non-write combined memory.

which seems vague enough to be implementation-defined .

However, I don't understand how `properties` should be used.

I was expecting `properties` to point to a value of type `cl_mem_alloc_flags_intel`, which could have the possible values `CL_MEM_ALLOC_DEFAULT_INTEL` or `CL_MEM_ALLOC_WRITE_COMBINED_INTEL`.

Instead it points to a value of  type  `pi_usm_mem_properties`, which is a `typedef` for `_pi_usm_mem_properties`, which is a `typedef` for
```
enum : pi_uint64 {
  PI_MEM_ALLOC_FLAGS = CL_MEM_ALLOC_FLAGS_INTEL
} _pi_usm_mem_properties;
```
, where `CL_MEM_ALLOC_FLAGS_INTEL` is `#define`d as `0x4195`.

@jbrodman I see that you authored most of [cl_intel_unified_shared_memory.asciidoc](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/USM/cl_intel_unified_shared_memory.asciidoc); could you clarify ?

Reply:
Sorry, meant add @bashbaug to this conversation.

Reply:
We probably need to fill in some more PI values.  I hadn't captured them all when I did the PI USM rewrite.

Reply:
At least on the OpenCL side, the `properties` are intended to provide extra information for a USM allocation.  Right now, the only valid property name is `CL_MEM_ALLOC_FLAGS_INTEL` to specify a property value bitfield `cl_mem_alloc_flags_intel`, which means they are really only useful to request write combined memory.

We have had a request to support traditional [CL_MEM_FLAGS](https://www.khronos.org/registry/OpenCL/specs/2.2/html/OpenCL_API.html#CL_MEM_FLAGS), to provide information such as whether an allocation is read-only, or inaccessible from the host, so don't be too surprised if that shows up in the future.

Note that all of these properties are optional - they give the runtime more information about an allocation that could improve performance, but they aren't required for functional correctness.

Reply:
> Alternatively, you could make this cuMemHostAlloc instead. We could map later the `CL_MEM_ALLOC_WRITE_COMBINED_INTEL` flag to `CU_MEMHOSTALLOC_WRITECOMBINED` assuming they are equivalent enough (they might not be)

Allocating write-combined memory in CUDA has the disadvantage that doesn't allow using the same pointer on the host and the device any more. From [5.13. Unified Addressing](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__UNIFIED.html#group__CUDA__UNIFIED):
> ### Exceptions, Disjoint Addressing
>
>Not all memory may be accessed on devices through the same pointer value through which they are accessed on the host. These exceptions are host memory registered using `cuMemHostRegister()` and host memory allocated using the flag `CU_MEMHOSTALLOC_WRITECOMBINED`. For these exceptions, there exists a distinct host and device address for the memory. The device address is guaranteed to not overlap any valid host pointer range and is guaranteed to have the same value across all contexts that support unified addressing. 

Unless there is a mechanism to handle distinct host and device pointers in the SYCL runtime or in the USM extension, I don't think implementing write-combined memory is a good idea.

Reply:
> Should there be an assert for properties to be nullptr since its not implemented?

I have to learn how properties work :-)

Reply:
OK, understood.

Currently, all calls to `piextUSMHostAlloc`, `piextUSMDeviceAlloc` and `piextUSMSharedAlloc` in [sycl/source/detail/usm/usm_impl.cpp](https://github.com/intel/llvm/blob/dc729a72b5bd815ffd7e118cd8d1992c5248ee4a/sycl/source/detail/usm/usm_impl.cpp) pass `nullptr` in the properties, so I've added an `assert` in that regard.